File: internal/driver/solace/clients.go
package solace

import (
	"context"
	"sort"
	"strconv"
	"strings"

	"github.com/amigoer/mq-studio/internal/model"
)

/*
 * Who is connected, and what they are doing.
 *
 * SEMP lists a Message VPN's clients with everything worth knowing about each
 * one, so this is one request. Two things about the list are worth saying out
 * loud because neither is obvious from the field names.
 *
 * The broker's own machinery appears here. A client whose name starts with "#"
 * is the broker talking to itself - the internal message bus, the REST
 * listener's own session, the MQTT bridge - and they are kept rather than
 * filtered, because a page that hid them would be hiding real connections
 * holding real resources. They are marked instead.
 *
 * There is no channel layer. A Solace client has flows, one per endpoint it is
 * bound to, and a flow is not a channel: it carries no prefetch of its own, it
 * is not something an operator closes, and it belongs to the endpoint as much
 * as to the client - which is why the queues page reports the bound count and
 * this one does not pretend to a multiplexing layer the protocol has not got.
 */

// clientRow is the shape the client collection answers with.
type clientRow struct {
	ClientName        string  `json:"clientName"`
	ClientAddress     string  `json:"clientAddress"`
	ClientUsername    string  `json:"clientUsername"`
	MsgVpnName        string  `json:"msgVpnName"`
	Description       string  `json:"description"`
	Platform          string  `json:"platform"`
	SoftwareVersion   string  `json:"softwareVersion"`
	Uptime            int64   `json:"uptime"`
	VirtualRouter     string  `json:"virtualRouter"`
	ClientProfileName string  `json:"clientProfileName"`
	ACLProfileName    string  `json:"aclProfileName"`
	SlowSubscriber    bool    `json:"slowSubscriber"`
	TLSDowngraded     bool    `json:"tlsDowngradedToPlainText"`
	DataRxByteCount   int64   `json:"dataRxByteCount"`
	DataTxByteCount   int64   `json:"dataTxByteCount"`
	RxByteRate        float64 `json:"rxByteRate"`
	TxByteRate        float64 `json:"txByteRate"`
	BindSuccessCount  int64   `json:"bindSuccessCount"`
}

const clientFields = "clientName,clientAddress,clientUsername,msgVpnName,description,platform," +
	"softwareVersion,uptime,virtualRouter,clientProfileName,aclProfileName,slowSubscriber," +
	"tlsDowngradedToPlainText,dataRxByteCount,dataTxByteCount,rxByteRate,txByteRate," +
	"bindSuccessCount"

// ListClientConnections lists what is connected to this Message VPN.
func (c *Conn) ListClientConnections(ctx context.Context, _ string) ([]*model.ClientConnection, error) {
	if err := c.live(); err != nil {
		return nil, err
	}
	// The namespace argument is ignored: a connection is one Message VPN, and
	// a listing that took another one would be reading outside its own scope.
	rows, err := listMonitor[clientRow](ctx, c.semp,
		"/msgVpns/"+segment(c.vpn)+"/clients?select="+clientFields)
	if err != nil {
		return nil, err
	}

	clients := make([]*model.ClientConnection, 0, len(rows))
	for _, row := range rows {
		host, port := splitAddress(row.ClientAddress)
		clients = append(clients, &model.ClientConnection{
			// The client name is the key rather than the address: one host can
			// hold twenty connections and the broker addresses each of them by
			// the name it chose or was given.
			Name:       row.ClientName,
			ClientName: row.ClientName,
			Namespace:  row.MsgVpnName,
			User:       row.ClientUsername,
			Node:       row.VirtualRouter,
			PeerHost:   host,
			PeerPort:   port,
			// The broker reports no protocol on a client, and there is no
			// honest way to derive one: an SMF session, an MQTT session and a
			// REST request all appear here with the same fields. The platform
			// string is what a reader gets instead, on the attributes.
			Protocol: "",
			State:    clientState(row),
			// Flows rather than channels, and they are not the same thing -
			// see the file comment. The count is the successful binds, which
			// is what the broker keeps.
			Channels:     int(row.BindSuccessCount),
			TLS:          false,
			RecvBytes:    row.DataRxByteCount,
			SendBytes:    row.DataTxByteCount,
			RecvByteRate: row.RxByteRate,
			SendByteRate: row.TxByteRate,
			Attributes: map[string]string{
				AttrClientPlatform: row.Platform,
				AttrClientVersion:  row.SoftwareVersion,
				AttrClientProfile:  row.ClientProfileName,
				AttrACLProfile:     row.ACLProfileName,
				AttrDescription:    row.Description,
				AttrUptimeSec:      strconv.FormatInt(row.Uptime, 10),
				AttrSlowSubscriber: strconv.FormatBool(row.SlowSubscriber),
				AttrTLSDowngraded:  strconv.FormatBool(row.TLSDowngraded),
				AttrInternal:       strconv.FormatBool(internalClient(row.ClientName)),
			},
		})
	}

	sort.Slice(clients, func(i, j int) bool { return clients[i].Name < clients[j].Name })
	return clients, nil
}

// Client attribute keys this driver fills in.
//
// A contract between this package and frontend/src/mq/solace/clients.ts, not
// part of the shared vocabulary.
const (
	AttrClientPlatform = "platform"
	AttrClientVersion  = "softwareVersion"
	AttrClientProfile  = "clientProfile"
	AttrACLProfile     = "aclProfile"
	AttrDescription    = "description"
	AttrUptimeSec      = "uptimeSec"
	AttrSlowSubscriber = "slowSubscriber"
	AttrTLSDowngraded  = "tlsDowngraded"
	// AttrInternal marks the broker talking to itself. Those connections are
	// listed rather than hidden - they hold real resources - and a reader
	// counting applications needs to be able to leave them out.
	AttrInternal = "internal"
)

/*
 * ListClientChannels is empty, and deliberately.
 *
 * A channel is AMQP 0-9-1's multiplexing layer and this protocol has not got
 * one. The nearest thing is a flow, and a flow is not a channel: it exists per
 * endpoint the client is bound to rather than per session, it carries no
 * prefetch of its own, and it is not something an operator closes. It is
 * reported as a bind count on the connection and as the bound count on the
 * queue, which is where it belongs.
 */
func (c *Conn) ListClientChannels(_ context.Context, _ string) ([]*model.ClientChannel, error) {
	return nil, nil
}

// clientState is the one word the broker's fields add up to.
//
// There is no state field on a client: it is in the list, so it is connected.
// What can be said beyond that is whether the broker has marked it as falling
// behind, which is the one condition worth seeing at a glance.
func clientState(row clientRow) string {
	if row.SlowSubscriber {
		return "slow"
	}
	return "connected"
}

// internalClient reports the broker's own machinery. Solace reserves a leading
// "#" for its own objects, and its internal sessions are named that way.
func internalClient(name string) bool {
	return strings.HasPrefix(name, "#")
}

// splitAddress takes the client's "host:port" apart. An address without a port
// keeps the whole string as the host rather than losing it.
func splitAddress(address string) (string, int) {
	colon := strings.LastIndexByte(address, ':')
	if colon < 0 {
		return address, 0
	}
	port, err := strconv.Atoi(address[colon+1:])
	if err != nil {
		return address, 0
	}
	return address[:colon], port
}

File: frontend/src/mq/solace/clients.ts
/**
 * Solace's view of a connected client.
 *
 * The keys are a contract with internal/driver/solace/clients.go.
 *
 * Two fields carry more than their names suggest. `internal` marks the broker
 * talking to itself - the message bus, the REST listener's own session - which
 * is listed rather than hidden because those hold real resources, and marked
 * because a reader counting applications has to be able to leave them out.
 * `slowSubscriber` is the broker's own verdict rather than a threshold this
 * app applied: it is set when a client cannot keep up with what is being
 * pushed to it, and it is the field that explains an egress rate that has
 * stopped climbing.
 */
import type { ClientConnection } from "@bindings/model/models";

const AttrPlatform = "platform";
const AttrSoftwareVersion = "softwareVersion";
const AttrClientProfile = "clientProfile";
const AttrACLProfile = "aclProfile";
const AttrDescription = "description";
const AttrUptimeSec = "uptimeSec";
const AttrSlowSubscriber = "slowSubscriber";
const AttrTLSDowngraded = "tlsDowngraded";
const AttrInternal = "internal";

export interface SolaceClient {
  /** The broker's own name for the session, and the key. */
  name: string;
  address: string;
  username: string;
  msgVpn: string;
  /** The client library and the machine it is running on, as it reported them. */
  platform: string | null;
  softwareVersion: string | null;
  clientProfile: string | null;
  aclProfile: string | null;
  description: string | null;
  uptimeSec: number | null;
  /** The broker's verdict that this client cannot keep up. */
  slowSubscriber: boolean;
  /** A TLS session the broker allowed to fall back to plain text. */
  tlsDowngraded: boolean;
  /** The broker talking to itself rather than an application. */
  internal: boolean;
  /** Endpoints it has bound to, which are flows rather than channels. */
  binds: number;
  recvBytes: number;
  sendBytes: number;
}

function attribute(client: ClientConnection, key: string): string | null {
  const value = client.attributes?.[key];
  return value == null || value === "" ? null : value;
}

function flag(client: ClientConnection, key: string): boolean {
  return client.attributes?.[key] === "true";
}

export function client(row: ClientConnection): SolaceClient {
  const uptime = attribute(row, AttrUptimeSec);
  const parsed = uptime == null ? Number.NaN : Number(uptime);
  return {
    name: row.name,
    address: row.peerPort > 0 ? `${row.peerHost}:${row.peerPort}` : row.peerHost,
    username: row.user,
    msgVpn: row.namespace,
    platform: attribute(row, AttrPlatform),
    softwareVersion: attribute(row, AttrSoftwareVersion),
    clientProfile: attribute(row, AttrClientProfile),
    aclProfile: attribute(row, AttrACLProfile),
    description: attribute(row, AttrDescription),
    uptimeSec: Number.isFinite(parsed) ? parsed : null,
    slowSubscriber: flag(row, AttrSlowSubscriber),
    tlsDowngraded: flag(row, AttrTLSDowngraded),
    internal: flag(row, AttrInternal),
    binds: row.channels,
    recvBytes: row.recvBytes,
    sendBytes: row.sendBytes,
  };
}

/** The applications, which is the list without the broker's own sessions. */
export function applications(clients: readonly SolaceClient[]): SolaceClient[] {
  return clients.filter((entry) => !entry.internal);
}
